pkg/qrcode/qrcode.go

```diff
@@ -0,0 +1,48 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package qrcode implements working with QR codes.
+package qrcode
+
+import (
+	"encoding"
+
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/types"
+)
+
+// Data represents QR code data.
+type Data interface {
+	encoding.TextMarshaler
+	encoding.TextUnmarshaler
+}
+
+// AuthenticatedEndDeviceIdentifiers defines end device identifiers with authentication code.
+type AuthenticatedEndDeviceIdentifiers interface {
+	AuthenticatedEndDeviceIdentifiers() (joinEUI, devEUI types.EUI64, authenticationCode []byte)
+}
+
+var errFormat = errors.DefineInvalidArgument("format", "invalid format")
+
+// Parse attempts to parse the given QR code data.
+func Parse(data []byte) (Data, error) {
+	for _, model := range [...]Data{
+		&LoRaAllianceTR005Draft2{},
```

Comment:
We might change this to a registry of formats, but then the order becomes important.

TR005 is easily recognizable by the header, but many QR codes are not, so switching on the prefix here is not very future proof.

Suggesting to figure out a more decent solution when we have more QR codes to deal with.